workitems/tokenlist/l2-token-list-v1.0-psd01.md

```diff
@@ -236,29 +239,43 @@ The tokens data element is a composite which MUST minimally contain the followin
 
 Note, that the detailed definition of the data elements in [[R2]](#r2) along with descriptions and examples are given in the schema itself below.
 
+[[R2]](#r2) testability: See suggested test fixtures for this documents' data schema below.
+
 #### **[D1]**
 All other data elements of the schema SHOULD be included in a representation of a canonical token list.
 
+[[D1]](#d1) testability: See suggested test fixtures for this documents' data schema below.
+
 #### **[CR1]>[D1]**
-if the extension data elements is used, the following data elements MUST be present in the schema representation:
+If the extension data elements is used, the following data elements MUST be present in the schema representation:
 - rootChainId
 - rootChainURI
 - rootAddress
 
 Note, that the detailed definition of the data elements in [[D1]](#d1) and [[CR1]>[D1]](#cr1d1) along with descriptions and examples are given in the schema itself below.
 
+[[CR1]>[D1]](#cr1d1) testability: See suggested test fixtures for this documents' data schema below.
+
 #### **[R3]**
 All properties in the schema identified in the description to be a Universal Resource Identifier (URI) MUST follow in their semantics [RFC 3986](#rfc3986).
 
+[[R3]](#r3) testability: All requirements for [RFC 3986](#rfc3986) are testable.
+
 #### **[R4]**
 The `chainId` property MUST follow [EIP-155](#eip155) standard.
 
+[[R4]](#r4) testability: Given that there are production implementations of [EIP-155](#eip155) such as Ethereum, all requirements of the standard are testable, and, therefore, by extension also [[R4]](#r4). 
```

Comment:
This rests  a false assumption that an implementation shows the specification is necessarily testable.

It *is* possible in principle to test that another implementation is compatible with the first, but in practice that requires specifying what actually needs to be interoperable.

Currently, neither this specification nor EIP-155 do so.

Reply:
@chaals since there is no certification program for EIPs and no official reference implementations, actual implementations that work are the only thing to go on. It is not in the remit of the L2 group to cure the existing standard ills of the Ethereum Ecosystem. If you want to take a hard line then no EIP can be referenced in an EEA Communities Project specification. However, that would be enormously counterproductive since it would discourage participation by Ethereum ecosystem participants since EIPs are the "gospel" atm. Hence, I would argue that the above is "good enough: for now. Once, the ecosystem has matured, we can most certainly revisit this point. Agreed?

Reply:
@chaals good to close this out?

Reply:
Here I think the testability explanation is papering over some assumptions. There doesn't need to be a formal certification system for EIP-155. It states two testable requirements, it should be possible in theory to test whether they are met, and I believe in practice it is easy and reasonable to do so.

However this is not implied by the existence of implementations. The testability rests on the fact that there are stated requirements and it is possible to test whether an implementation meets them.

I don't think it's OK to include assumptions based on what a spec implies you need to do as part of what you expect has been tested, because it is too often too easy to find people whose assumptions are different.

I think it's easy to state the EIP-155 requirements that are testable and (by D3/R5) need to be tested:
9 elements are hashed not 6, and the `v` of the signature is set as described.

But that isn't something that applies to the `chainID`, it's about how transactions are hashed on the chain. So the requirement should be tweaked to align with that.

IMHO it makes sense for the use case of this spec to be slightly stricter than EIP-155, and *require* that there is a `chainID` and that 9 elements are hashed as described - instead of leaving open the ambiguity that people can read into "**should**", possibly making specific exceptions for things that are so old they didn't do this, rather than leaving in a generically available "waiver" clause.

Reply:
@chaals how about the following?

#### **[R4]**
The `chainId` property utilized MUST allow for the requirements of the [EIP-155](#eip155) standard to be met.

Namely, transaction replay protection on the network that is identified by the `chainId` property value. Note, that for replay protection to be guaranteed, the `chainId` should be unique. Ensuring a unique `chainId` is beyond the scope of this document.

[[R4]](#r4) testability:  [EIP-155](#eip155) requires that a transaction hash is derived from the `keccak256` hash of the following nine RLP encoded elements `(nonce, gasprice, startgas, to, value, data, chainid, 0, 0)` which can be tested easily with existing cryptographic libraries. [EIP-155](#eip155) further requires that the `v` value of the `secp256k1` signature must be set to `{0,1} + CHAIN_ID * 2 + 35` where `{0,1}` is the parity of the `y` value of the curve point for which the signature `r`-value is the `x`-value in the secp256k1 signing process. This requirement is testable with available open-source secp256k1 digital signature suites. Therefore, [[R4]](#r4) is testable.

Reply:
Any `chainID` allows for the requirements to be met.

I think the requirement R4 should state something like 

> (relevant?) transactions must be hashed with 9 RLP elements per EIP-155 (except for MainNet before block 2675000[, other exceptions]?), and implement the EIP-155 requirement for the `v` value of the signature.

Testability is then straightforward - either transactions are or are not hashed appropriately.

I think it makes sense to add a requirement that tries to make for unique chainID values. If EIP-3220 were done I would suggest that, and I think it's worth allowing for and anticipating in spec prose the possibility to update this spec if EIP-3220 does get done - which means there is a reasonable discussion to be had about whether the current SHOULD is appropriate, although I personally think it's OK. Something like `chainID`s must be registered on the github registry EIP-155 points to and not conflict with any other chain registered there.

Yes, the Github registry can change or vanish. Until it does it makes sense to use it, and doing so adds to the pressure for it not to fail... But for that matter the entire EIP record can *in principle* disappear. For now, we rely on our faith that the safeguards against that are sufficiently strong. Without that we couldn't build anything...

Reply:
@chaals 





> Any `chainID` allows for the requirements to be met.

That is not correct, as the proposed note to [R4] states. For ensuring full replay protection, not only on a specific chain, a unique `chainId` is not a sufficient condition. Also, a cross-chain problem only exists for forked chains, and then only for contracts that existed before the fork. Hence, for only a very small number of L1 chains, certainly not L2s (at least I have not heard of a (state) forked L2. 

I believe that the suggestion as written is more than sufficient.

Reply:
Sorry, I mean any `chainID` can meet the requirements of EIP-155 - because the clearly expressed requirements are just that you hash the `chainID` with some other stuff.

I haven't heard of a state forked L2 yet either. But it seems reasonably easy to accomplish in practice, and assuming that it won't happen as a basis for security doesn't seem wise to me.

Reply:
@chaals
- It would be correct to say "any `chainId` that is implemented in a network protocol" since a transaction hash generated with a `chainId` that does not match the `chainId` of the target network would be rejected by the transaction's target network. In addition, the nonce of the sender account and its nonce would have to be the same as on the spoof chain, and the `to` address would have to exist and the `data` would have to trigger the same function on the spoof chain as on the target chain.
- Furthermore, for this to work on an L2, the above would have to correct, AND the same bridge contract at the time of the fork would have to be deployed, AND the bridge contract data would have to be replayed one by one to get to the same bridge contract state, and it must happen on the SAME L1

I do not know how you want to replay all the L2 transactions on the L1 because the nonces and senders do not match on the L1 anymore, so a replay on the SAME L1 is not possible and a different L1 does not work because then the token is provably different than the one from the original L1 since the `chainIds` are different! If you use a fork of the L1 you still have the problem that you cannot reply past transactions that already happened on the forked L1 as you would have to because you need to get the bridge contract state properly updated in the spoof bridge contract on the L1. 

Did I miss anything?

Reply:
@chaals The WG during its 1/25 meeting discussed the uniqueness point you raised. It was decided that we will not add a uniqueness requirement. A security advisory was added in the security considerations that implementers of the canonical token list standard must ensure the correctness and security of the data they display. Happy to discuss this further on a topic and possibly create an EIP for `chainId` uniqueness. Will resolve this conversation for now.